modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java

```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
+                    // in this case as it shouldn't cause upstream request to fail.
+                    logger.warn("Exception occurred while putting item to disk cache", ex);
+                    exceptionOccurredOnDiskCachePut = true;
+                }
+                if (!exceptionOccurredOnDiskCachePut) {
+                    updateStatsOnPut(TIER_DIMENSION_VALUE_DISK, key, notification.getValue());
+                }
             }
-        } else {
-            try {
-                value = future.get().v2();
-            } catch (InterruptedException ex) {
-                throw new IllegalStateException(ex);
+            if (!canCacheOnDisk || exceptionOccurredOnDiskCachePut) {
+                // If the value is not going to the disk cache, send this notification to the TSC's removal listener
+                // as the value is leaving the TSC entirely
+                removalListener.onRemoval(notification);
+                countEvictionTowardsTotal = true;
             }
+            updateStatsOnRemoval(TIER_DIMENSION_VALUE_ON_HEAP, wasEvicted, key, notification.getValue(), countEvictionTowardsTotal);
         }
-        return value;
-    }
 
-    @Override
-    public void invalidate(ICacheKey<K> key) {
-        // We are trying to invalidate the key from all caches though it would be present in only of them.
-        // Doing this as we don't know where it is located. We could do a get from both and check that, but what will
-        // also trigger a hit/miss listener event, so ignoring it for now.
-        // We don't update stats here, as this is handled by the removal listeners for the tiers.
-        for (Map.Entry<ICache<K, V>, TierInfo> cacheEntry : caches.entrySet()) {
-            if (key.getDropStatsForDimensions()) {
-                List<String> dimensionValues = statsHolder.getDimensionsWithTierValue(key.dimensions, cacheEntry.getValue().tierName);
-                statsHolder.removeDimensions(dimensionValues);
+        boolean evaluatePolicies(V value) {
+            for (Predicate<V> policy : policies) {
+                if (!policy.test(value)) {
+                    return false;
+                }
             }
-            if (key.key != null) {
-                try (ReleasableLock ignore = writeLock.acquire()) {
-                    cacheEntry.getKey().invalidate(key);
+            return true;
+        }
+
+        /**
+         * Get a value from the tiered cache, and the name of the tier it was found in.
+         * @param captureStats Whether to record hits/misses for this call of the function
+         * @return A tuple of the value and the name of the tier it was found in.
+         */
+        private Function<ICacheKey<K>, Tuple<V, String>> getValueFromTieredCache(boolean captureStats, boolean forceCheck) {
+            return key -> {
+                try (ReleasableLock ignore = readLock.acquire()) {
```

Comment:
Added this new forceCheck parameter without updating the documentation. We should avoid such changes as part of large refactoring. Can you explain the purpose of this new parameter?

Reply:
My bad, seems like this change is not needed and made it here due to some sample changes I was trying. Will remove.